DB/src/main/java/io/deephaven/db/v2/sources/regioned/RegionedColumnSourceManager.java

```diff
@@ -241,69 +248,97 @@ public ImmutableTableLocationKey getKey(@NotNull final EmptyTableLocationEntry e
+                    //noinspection unchecked
+                    final ColumnLocationState state = new ColumnLocationState(
+                            columnDefinition,
+                            columnSources.get(columnDefinition.getName()),
+                            location.getColumnLocation(columnDefinition.getName()));
+                    columnLocationStates.add(state);
+                    state.regionAllocated(regionIndex);
+                    if (state.needToUpdateGrouping()) {
+                        state.updateGrouping(addIndexInTable == null ? addIndexInTable = initialIndex.shift(regionFirstKey) : addIndexInTable);
+                    }
+                }
+            } finally {
+                if (addIndexInTable != null) {
+                    addIndexInTable.close();
+                }
             }
-            sizeAtLastUpdate = size;
+            indexAtLastUpdate = initialIndex;
         }
 
-        private void pollSizeUpdates(final Index.SequentialBuilder addedIndexBuilder) {
+        private void pollUpdates(final Index.SequentialBuilder addedIndexBuilder) {
             Assert.neqNull(subscriptionBuffer, "subscriptionBuffer"); // Effectively, this is asserting "isRefreshing".
             if (!subscriptionBuffer.processPending()) {
                 return;
             }
-            final long size = location.getSize();
-            if (size == TableLocationState.NULL_SIZE) {
-                // This should be impossible - the subscription buffer transforms a transition to NULL_SIZE into a pending exception
-                throw new TableDataException("Location " + location + " is no longer available, data has been removed");
-            }
-            if (size < sizeAtLastUpdate) { // Bad change
-                throw new IllegalStateException("Size decreased for location " + location + ": was " + sizeAtLastUpdate + ", now " + size);
-            }
-            if (size == sizeAtLastUpdate) {
-                // Nothing to do
-                return;
-            }
-            if (size > RegionedColumnSource.REGION_CAPACITY_IN_ELEMENTS) {
-                throw new TableDataException("Location " + location + " has updated size " + size
-                        + ", larger than maximum supported location size " + RegionedColumnSource.REGION_CAPACITY_IN_ELEMENTS);
-            }
+            final ReadOnlyIndex updateIndex = location.getIndex();
+            try {
+                if (updateIndex == null) {
+                    // This should be impossible - the subscription buffer transforms a transition to null into a pending exception
+                    throw new TableDataException("Location " + location + " is no longer available, data has been removed");
+                }
+                if (!indexAtLastUpdate.subsetOf(updateIndex)) { // Bad change
+                    throw new IllegalStateException("Index keys removed at location " + location + ": " + indexAtLastUpdate.minus(updateIndex));
```

Comment:
Is an IllegalStateException better than an Assert here?
The behavior on the grpc_api main for IllegalStateExceptions, if I remember correctly, is not great.
Also, maybe we need tickets to ensure that, on asserts failed, we get a heap dump, and on IllegalStateException,
we don't leave the worker in a foobar state.

Reply:
This could arguably happen with a bad publisher, it doesn't have to be an engine bug. I tend to think for possibly-user-facing errors we should use exceptions with clean messages, rather than assert. I'm sensitive to your heap dump point, though, and so I'm planning to change to:
```
                    Assert.statementNeverExecuted("Index keys removed at location " + location + ": " + indexAtLastUpdate.minus(updateIndex));
```
The comment regarding `grpc_api` main is not something I intend to worry about at this layer; if we encounter an error we throw an exception. The calling code needs to be wrapped in a try-catch block in order to surface appropriate errors to the user. We shouldn't crash just because an engine op failed, unless we setup failure listeners to make that happen deliberately.